jujugui/static/gui/src/app/test-init.js

```diff
@@ -54,17 +55,582 @@ describe('init', () => {
+        // IE10 and 11 do not have the dataTransfer.items property during hover
+        // so we cannot tell what type of file is being hovered over the canvas.
+        // So we will just return the default which is "yaml".
+        const result = app._determineFileType({
+          types: ['Files']
+        });
+        assert.equal(result, '');
+      });
+    });
+
+    describe('UI notifications', () => {
+      it('_renderDragOverNotification renders drop UI', () => {
+        const fade = sinon.stub();
+        const reactdom = sinon.stub(ReactDOM, 'render');
+        cleanups.push(reactdom.restore);
+        app.topology.fadeHelpIndicator = fade;
+        app._renderDragOverNotification();
+        assert.equal(fade.callCount, 1);
+        assert.equal(fade.lastCall.args[0], true);
+        assert.equal(reactdom.callCount, 1);
+      });
+
+      it('_hideDragOverNotification hides drop UI', () => {
+        const fade = sinon.stub();
+        const reactdom = sinon.stub(
+          ReactDOM, 'unmountComponentAtNode');
+        cleanups.push(reactdom.restore);
+        app.topology.fadeHelpIndicator = fade;
+        app._hideDragOverNotification();
+        assert.equal(fade.callCount, 1);
+        assert.equal(fade.lastCall.args[0], false);
+        assert.equal(reactdom.callCount, 1);
+      });
+    });
+
+    it('dispatches drag events properly: _appDragOverHandler', () => {
+      const determineFileTypeStub = sinon.stub(
+        app, '_determineFileType').returns('zip');
+      const renderDragOverStub = sinon.stub(
+        app, '_renderDragOverNotification');
+      const dragTimerControlStub = sinon.stub(
+        app, '_dragleaveTimerControl');
+      cleanups = cleanups.concat([
+        determineFileTypeStub.restore,
+        renderDragOverStub.restore,
+        dragTimerControlStub
+      ]);
+
+      const noop = () => {};
+      const ev1 = {
+        dataTransfer: 'foo', preventDefault: noop, type: 'dragenter' };
+      const ev2 = { dataTransfer: {}, preventDefault: noop, type: 'dragleave' };
+      const ev3 = { dataTransfer: {}, preventDefault: noop, type: 'dragover' };
+
+      app._appDragOverHandler(ev1);
+      app._appDragOverHandler(ev2);
+      app._appDragOverHandler(ev3);
+
+      assert.equal(determineFileTypeStub.callCount, 3);
+      assert.equal(renderDragOverStub.calledOnce, true);
+      assert.equal(dragTimerControlStub.callCount, 3);
+      const args = dragTimerControlStub.args;
+      assert.equal(args[0][0], 'start');
+      assert.equal(args[1][0], 'start');
+      assert.equal(args[2][0], 'stop');
+    });
+
+    it('can start and stop the drag timer: _dragLeaveTimerControl', () => {
+      app._dragleaveTimerControl('start');
+      assert.equal(app._dragLeaveTimer !== undefined, true);
+      app._dragleaveTimerControl('stop');
+      assert.equal(app._dragLeaveTimer === null, true);
+    });
+  });
+
+  xdescribe('Application authentication', () => {
```

Comment:
There are a couple of tests in here that are failing. @hatched if you had a minute to have a look that'd be great.

Reply:
I've figured these out now.